File: leetcode/445_add_two_numbers_ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        l1 = self.reverse(l1)
        l2 = self.reverse(l2)
        return self.reverse(self.addTwoNumbers_recursive(l1, l2, 0))
    
    def reverse(self, head: ListNode):
        if head == None or head.next == None:
            return head
        p = self.reverse(head.next)
        head.next.next = head
        head.next = None
        return p
 
 
    def addTwoNumbers_recursive(self, l1: ListNode, l2: ListNode, carry: int):
        if l1 == None and l2 == None:
            if carry > 0:
                return ListNode(carry)
            else:
                return None
        elif l1 == None:    
            curr_sum = l2.val + carry
        elif l2 == None:
            curr_sum = l1.val + carry
        else:
            curr_sum = l1.val + l2.val + carry
        node = ListNode(curr_sum % 10)
        node.next = self.addTwoNumbers_recursive(l1.next if l1 != None else l1, l2.next if l2 != None else l2, curr_sum // 10)
        return node
 

File: leetcode/445_add_two_numbers_ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        stack<int> s1;
        stack<int> s2;
        ListNode* head = l1;
        while (head != NULL) {
            s1.push(head->val);
            head = head -> next;
        }
        head = l2;
        while (head != NULL) {
            s2.push(head->val);
            head = head->next;
        }
        ListNode* dummy = NULL;
        int carry = 0;
        while(!s1.empty() || !s2.empty()) {
            int sum = carry;
            if (!s1.empty()) {
                sum += s1.top();
                s1.pop();
            }
            if (!s2.empty()){
                sum += s2.top();
                s2.pop();
            }
            ListNode* newNode = new ListNode(sum % 10, NULL);
            newNode->next = dummy;
            dummy = newNode;
            carry = sum / 10;
        }
        if (carry > 0) {
            ListNode* newNode = new ListNode(carry);
            newNode->next = dummy;
            return newNode;
        }
        return dummy;
    }
};
